sktime/utils/data_container.py

```diff
@@ -1,230 +1,245 @@
+
+    Parameters
+    ----------
+    X : pandas DataFrame
+    return_arrays : bool, optional (default=False)
+        - If True, returns a numpy arrays within cells of nested pandas DataFrame.
+        - If False, returns a pandas Series within cells.
+    index : array-like, shape=[n_samples], optional (default=None)
+        Sample (row) index of transformed dataframe
+    time_index : array-like, shape=[n_obs], optional (default=None)
+        Time series index of transformed dataframe
+
+    Returns
+    -------
+    Xt : pandas DataFrame
+        Transformed dataframe in nested format
+    """
+
+    if (time_index is not None) and return_arrays:
+        raise ValueError("`Time_index` cannot be specified when `return_arrays` is True, time index can only be set to "
+                         "pandas Series")
+
+    container = np.array if return_arrays else pd.Series
+
+    n_samples, n_obs = X.shape
+
+    if time_index is None:
+        time_index = np.arange(n_obs)
+    kwargs = {'index': time_index}
+
+    Xt = pd.DataFrame(pd.Series([container(X.iloc[i, :].values, **kwargs) for i in range(n_samples)]))
+
+    if index is not None:
+         Xt.index = index
+
+    return Xt
+
+
+tabularise = tabularize
+
+
+detabularise = detabularize
+
+
+def concat_nested_arrays(arrs, return_arrays=False):
+    """
+    Helper function to nest tabular arrays from nested list of arrays.
+
+    Parameters
+    ----------
+    arrs : list of numpy arrays
+        Arrays must have the same number of rows, but can have varying number of columns.
+    return_arrays: bool, optional (default=False)
+        - If True, return pandas DataFrame with nested numpy arrays.
+        - If False, return pandas DataFrame with nested pandas Series.
+
+    Returns
+    -------
+    Xt : pandas DataFrame
+        Transformed dataframe with nested column for each input array.
+    """
+    if return_arrays:
+        Xt = pd.DataFrame(np.column_stack(
+            [pd.Series([np.array(vals) for vals in interval])
+             for interval in arrs]))
+    else:
+        Xt = pd.DataFrame(np.column_stack(
+            [pd.Series([pd.Series(vals) for vals in interval])
+             for interval in arrs]))
+    return Xt
+
+
+def get_time_index(X):
+    """Helper function to get index of time series data
+
+    Parameters
+    ----------
+    X : pandas DataFrame
+
+    Returns
+    -------
+    time_index : pandas Index
+        Index of time series
+    """
+
+    # assumes that all samples share the same the time index, only looks at first row
+    if isinstance(X, pd.DataFrame):
+        Xs = X.iloc[0, 0]
+
+    elif isinstance(X, pd.Series):
+        Xs = X.iloc[0]
+
+    else:
+        raise ValueError(f"X must be a pandas DataFrame or Series, but found: {type(X)}")
+
+    # get time index
+    time_index = Xs.index if hasattr(Xs, 'index') else pd.RangeIndex(Xs.shape[0])
+
+    return time_index
+
+def dataframe_to_numpy(X, a = None, b = None):

```

Comment:
Thanks for the utility function, could we rename it to `nested_to_3d_numpy`? This is very useful, as it also provides an interface to the tslearn package 👍

Reply:
Renamed to `nested_to_3d_numpy`.